### PR TITLE
zephyr: ccp: Added support for testing Call Control client.

### PR DESCRIPTION
**Pull request by MSEEHenrik**
- Discover GTBS.
- Accept call.
- Terminate call.
- Originate call.
- Get call states.

### CODE REVIEW

File: autopts/pybtp/btp/ccp.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+    success = ccp_await_event('Discovered', timeout)
+
+    return success, \
+           events['Discovered']['status'], \
+           events['Discovered']['tbs_count'], \
+           events['Discovered']['gtbs']
+
+
+def ccp_ev_discovered(data, data_len):
```
Comment (mkasenberg) [inline autopts/pybtp/btp/ccp.py:191]:
```suggestion
def ccp_ev_discovered(ccp, data, data_len):
```

File: autopts/pybtp/btp/ccp.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                             CONTROLLER_INDEX),
+    'read_call_state':     ( defs.BTP_SERVICE_ID_CCP,
+                             defs.CCP_READ_CALL_STATE,
+                             CONTROLLER_INDEX)
+}
+
+events = {
+    'Discovered': { 'count': 0, 'status': 0, 'tbs_count': 0, 'gtbs': False },
+    'Call_State': { 'count': 0, 'status': 0, 'index': 0, 'call_count': 0, 'states': [] }
+}
```
Comment (mkasenberg) [inline autopts/pybtp/btp/ccp.py:55]: Maybe events should be implemented as a part of Stack class, so they could be accessed easily across layers
https://github.com/auto-pts/auto-pts/blob/263c5c33d279eef423a0d23d512f9bf01d3d1e86/autopts/ptsprojects/stack.py#L1353-L1369

Like it has been done for previous layers, e.g.:
https://github.com/auto-pts/auto-pts/blob/263c5c33d279eef423a0d23d512f9bf01d3d1e86/autopts/ptsprojects/stack.py#L704-L710

**Review by mkasenberg — CHANGES_REQUESTED** [review]

File: autopts/pybtp/btp/btp.py [hunk 3]
```diff
@@ -685,6 +695,12 @@ def event_handler(hdr, data):
             cb(stack_obj, data[0], hdr.data_len)
             return True
 
+    elif hdr.svc_id == defs.BTP_SERVICE_ID_CCP:
+        if hdr.op in CCP_EV:
+            cb = CCP_EV[hdr.op]
+            cb(data[0], hdr.data_len)
+            return True
+
```
[the remark this answers is not in the document]
Comment (MSEEHenrik) [reply autopts/pybtp/btp/btp.py]: Done.

File: autopts/pybtp/btp/ccp.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    if flags and not flags & CallFlags.OUTGOING:
+        result = '|' + result if result else result
+        result = 'INCOMING' + result
+    if not flags & CallFlags.WITHHELD_BY_NETWORK:
+        result = result + '|' if result else result
+        result = result + 'PROVIDED_BY_NETWORK'
+    return result
+
+
+def ccp_ev_call_states(data, data_len):
```
[the remark this answers is not in the document]
Comment (MSEEHenrik) [reply autopts/pybtp/btp/ccp.py]: Done.

File: autopts/pybtp/btp/ccp.py (hunk 1 above)
Comment (MSEEHenrik) [reply autopts/pybtp/btp/ccp.py]: Done.

File: autopts/pybtp/btp/ccp.py (hunk 2 above)
Comment (MSEEHenrik) [reply autopts/pybtp/btp/ccp.py]: Done.

File: autopts/pybtp/btp/ccp.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+
+def ccp_ev_discovered(data, data_len):
+    status, tbs_count, gtbs_found = struct.unpack('<IB?', data)
+    logging.debug(f"{ccp_ev_discovered.__name__} status: %u tbs count: %u gbts: %s" % (status, tbs_count, gtbs_found))
+
+    events['Discovered']['status'] = status
+    events['Discovered']['tbs_count'] = tbs_count
+    events['Discovered']['gtbs'] = gtbs_found
+    events['Discovered']['count'] += 1
```
[the remark this answers is not in the document]
Comment (MSEEHenrik) [reply autopts/pybtp/btp/ccp.py]: Done.

**Review by mkasenberg — APPROVED** [review]

File: doc/btp_ccp.txt [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+		Each bit in response is a flag indicating if command with
+		opcode matching bit number is supported. Bit set to 1 means
+		that command is supported. Bit 0 is reserved and shall always
+		be set to 0. If specific bit is not present in response (less
+		than required bytes received) it shall be assumed that command
+		is not supported.
+
+		In case of an error, the error response will be returned.
+
+	Opcode 0x02: CCP Register Callbacks
```
Comment (sjanc) [inline doc/btp_ccp.txt:23]: I'm not sure about this command.  This looks like some implementation detail of zephyr stack exposed over BTP...

Can you explain why is this needed?

**Review by sjanc — CHANGES_REQUESTED** [review]

File: doc/btp_ccp.txt (hunk 6 above)
Comment (MSEEHenrik) [reply doc/btp_ccp.txt:23]: Yes - This is part of the essential setup of the CCP.  It is used to register all callbacks from both discovery and individual call control commands.  Without it - there is no CCP functionality.
You could argue that it could have been called internally by including a check in each command, but that would just mean a lot of extra housekeeping.

Comment (sjanc) [reply doc/btp_ccp.txt:23]: but this is zephyr API specific, if you need to register any callbacks in tester app just do it on service init

Comment (MSEEHenrik) [reply doc/btp_ccp.txt:23]: Yes - sure that is possible.  I will do that.

Comment (MSEEHenrik) [reply doc/btp_ccp.txt:23]: Done.

**Review by sjanc — APPROVED** [review]

**Comment by sjanc** [comment]
@MSEEHenrik  looks good, but please rebase

**Comment by MSEEHenrik** [comment]
Thanks.  Sure, will rebase as soon as I get a window.  We are are in meetings for the next two days.

**Comment by sjanc** [comment]
@MSEEHenrik after second look,  I wonder if those command should have address to identify peer (like other client btp) ?

**Comment by MSEEHenrik** [comment]
We could do that, but does it add any value?  We always target the PTS dongle.  I can see that some services include the address and some don't.  It isn't really consistent.

**Comment by mkasenberg** [comment]
There are some test cases for some profiles where 2 PTS dongles are needed and IUT has to e.g. to discover MICP services of Lower Tester 2. Are there any of those in CCP/TBS test specs? They are sometimes included in the test specs for upper layer profiles.

**Comment by MSEEHenrik** [comment]
No - not to my knowledge.

**Comment by MSEEHenrik** [comment]
OK - No problem.  I will add the peer address to the BTP commands.

**Comment by MSEEHenrik** [comment]
Done.
